dm/master/coordinator/coordinator.go

```diff
@@ -55,9 +63,87 @@ func NewCoordinator() *Coordinator {
 	}
 }
 
-// Init would recover infomation from etcd
-func (c *Coordinator) Init(etcdClient *clientv3.Client) {
+// Start starts the coordinator and would recover infomation from etcd.
+func (c *Coordinator) Start(ctx context.Context, etcdClient *clientv3.Client) error {
 	// TODO: recover upstreams and configs and workers
+	// workers
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.etcdCli = etcdClient
+
+	// recovering.
+	ectx, cancel := context.WithTimeout(etcdClient.Ctx(), etcdTimeouit)
+	defer cancel()
+	resp, err := etcdClient.Get(ectx, common.WorkerRegisterKeyAdapter.Path(), clientv3.WithPrefix())
+	if err != nil {
+		return err
+	}
+
+	for _, kv := range resp.Kvs {
+		//addr := strings.TrimPrefix(common.WorkerRegisterKeyAdapter.Path(), string(kv.Key))
+		addr := common.WorkerRegisterKeyAdapter.Decode(string(kv.Key))[0]
+		name := string(kv.Value)
+		c.workers[addr] = NewWorker(name, addr)
+		log.L().Info("load worker successful", zap.String("addr", addr), zap.String("name", name))
+	}
+
+	// configs
+	resp, err = etcdClient.Get(ectx, common.UpstreamBoundWorkerKeyAdapter.Path(), clientv3.WithPrefix())
+	if err != nil {
+		return nil
+	}
+
+	for _, kv := range resp.Kvs {
+		addr := common.UpstreamBoundWorkerKeyAdapter.Decode(string(kv.Key))[0]
+		sourceID := string(kv.Value)
+		w, ok := c.workers[addr]
+		if !ok {
+			log.L().Error("worker not exist but binding relationship exist", zap.String("addr", addr), zap.String("source", sourceID))
+			continue
+		}
+		resp, err = etcdClient.Get(ctx, common.UpstreamConfigKeyAdapter.Encode(sourceID))
```

Comment:
Does this `resp` overwrite the value of it in L96?

Reply:
actually `for ... range` will clone the values, not affect the origin value.